### DISCUSSION TITLE
Make stdout.write() return number of bytes actually written?

### DISCUSSION THREAD

**Discussion by mvds00** [micropython/micropython#11807] [Ideas]
Not sure if this is a bug or a feature, but it seems that `sys.stdout.write()` returns the length of the data that should have been written, regardless of the number of bytes actually written. This is confusing behavior when dealing with data loss issues. I have seen no warnings about this type of behavior in the documentation.

I've been researching an issue where after a few hours of runtime, incidentally chunks are missing from the data stream sent out via USB, and I've narrowed it down to somewhere between main.py code `sys.stdout.write(buf)` and the receiving end (an ARM based Linux device with ACM driver).

It seems that in the MicroPython core code, this is how stdout data is sent, in `runtime/sys_stdio_mphal.c`:

```
STATIC mp_uint_t stdio_write(mp_obj_t self_in, const void *buf, mp_uint_t size, int *errcode) {
    sys_stdio_obj_t *self = MP_OBJ_TO_PTR(self_in);
    if (self->fd == STDIO_FD_OUT || self->fd == STDIO_FD_ERR) {
        mp_hal_stdout_tx_strn_cooked(buf, size);
        return size;
    } else {
        *errcode = MP_EPERM;
        return MP_STREAM_ERROR;
    }
}
```

This calls `mp_hal_stdout_tx_strn_cooked()` and returns the number of (uncooked) bytes that _should have been written_, no matter what.

Likewise, for `sys.stdout.buffer.write()`:

```
STATIC mp_uint_t stdio_buffer_write(mp_obj_t self_in, const void *buf, mp_uint_t size, int *errcode) {
    mp_hal_stdout_tx_strn(buf, size);
    return size;
}
```

From `mp_hal_stdout_tx_strn_cooked()`, `mp_hal_stdout_tx_strn()` is called as well, which is board specific, and in my case in `ports/rp2/mphalport.c`:

```
// Send string of given length
void mp_hal_stdout_tx_strn(const char *str, mp_uint_t len) {
    #if MICROPY_HW_ENABLE_UART_REPL
    mp_uart_write_strn(str, len);
    #endif

    #if MICROPY_HW_USB_CDC
    if (tud_cdc_connected()) {
        for (size_t i = 0; i < len;) {
            uint32_t n = len - i;
            if (n > CFG_TUD_CDC_EP_BUFSIZE) {
                n = CFG_TUD_CDC_EP_BUFSIZE;
            }
            int timeout = 0;
            // Wait with a max of USC_CDC_TIMEOUT ms
            while (n > tud_cdc_write_available() && timeout++ < MICROPY_HW_USB_CDC_TX_TIMEOUT) {
                MICROPY_EVENT_POLL_HOOK
            }
            if (timeout >= MICROPY_HW_USB_CDC_TX_TIMEOUT) {
                break;
            }
            uint32_t n2 = tud_cdc_write(str + i, n);
            tud_cdc_write_flush();
            i += n2;
        }
    }
    #endif

    #if MICROPY_PY_OS_DUPTERM
    mp_os_dupterm_tx_strn(str, len);
    #endif
}
```

This looks like it might be the source of my issue. (Or maybe also that `timeout` is counting more than just millis... but that's a different topic.)

Since I encountered this issue at the very start of my first MicroPython project, I suspect other people will run into this. So the idea I would like to share is to make `stdout.write()` behave like any other `write()` call, returning the number of bytes actually written. I am willing to write and submit the patch, but before doing so I'd like to know whether there are reasons why such behavior is not already implemented. One related issue I found was https://github.com/micropython/micropython/issues/7479 where someone is using USB stdout to transfer large amounts of data - so there is definitely a use case for high volume and reliable `sys.stdout.write()`.

**Comment by mvds00** [comment]
After extensive testing, I can confirm that making `sys.stdout.buffer.write()` return the actual number of bytes written, proves that:
- Short writes indeed caused the data loss issue I was experiencing, which is now solved for me.
- The fix is very simple with little impact (firmware.bin for ports/rp2 is 8 bytes larger).
- Measurement of time spent in `sys.stdout.buffer.write()` indicates that often about 500 ms are spent until timeout, but sometimes timeout happens after about 250 ms already, which is a sign that `timeout` is counting more than just millis.
- I now have a new issue to solve, as apparently USB blocks output for up to a second, irregularly but three times per minute on average from the moment the issue starts, often 1-2 hours after start at 3 kB/s average transfer speed (but this could also be an issue on the other end of the cable).

`sys.stdout.write()` is a bit more difficult (but not impossible) to fix, as it modifies (enlarges) data by "cooking" the string, replacing `\n` by `\r\n`.

Any feedback or thoughts are more than welcome.

**Comment by jimmo** [comment]
@thingsconnected MicroPython should match CPython's behavior, and I agree that the proposed change also makes sense. I think it sounds like it's worth submitting a PR. Thanks!

**Reply by mvds00** [reply]
It has been a while... I updated the PR and hope it can be merged.
